File: Divide-and-Conquer Searching/Binary Search/Binary Search On Answer/BinarySearchOnAnswer.c
#include <stdio.h>

int canComplete(int target, int hours[], int size, int maxHours)
{
    int totalHours = 0;

    for (int i = 0; i < size; i++)
    {
        totalHours += (hours[i] + target - 1) / target;

        if (totalHours > maxHours)
        {
            return 0;
        }
    }

    return 1;
}

int binarySearchOnAnswer(int hours[], int size, int maxHours)
{
    int left = 1;
    int right = hours[0];

    for (int i = 1; i < size; i++)
    {
        if (hours[i] > right)
        {
            right = hours[i];
        }
    }

    int answer = right;

    while (left <= right)
    {
        int mid = left + (right - left) / 2;

        if (canComplete(mid, hours, size, maxHours))
        {
            answer = mid;
            right = mid - 1;
        }
        else
        {
            left = mid + 1;
        }
    }

    return answer;
}

int main()
{
    int piles[] = {3, 6, 7, 11};
    int size = sizeof(piles) / sizeof(piles[0]);
    int h = 8;

    printf("Minimum Speed: %d\n",
           binarySearchOnAnswer(piles, size, h));

    return 0;
}

File: Divide-and-Conquer Searching/Binary Search/Binary Search On Answer/BinarySearchOnAnswer.py
def can_complete(speed, piles, hours):
    total_hours = 0

    for pile in piles:
        total_hours += (pile + speed - 1) // speed

        if total_hours > hours:
            return False

    return True


def binary_search_on_answer(piles, hours):
    left = 1
    right = max(piles)

    answer = right

    while left <= right:
        mid = left + (right - left) // 2

        if can_complete(mid, piles, hours):
            answer = mid
            right = mid - 1
        else:
            left = mid + 1

    return answer


if __name__ == "__main__":
    piles = [3, 6, 7, 11]
    hours = 8

    print(
        "Minimum Speed:",
        binary_search_on_answer(
            piles,
            hours
        )
    )
